nixos/modules/services/monitoring/prometheus/exporters.nix

```diff
@@ -183,6 +183,9 @@ let
         "ip46tables -A nixos-fw ${conf.firewallFilter} "
         "-m comment --comment ${name}-exporter -j nixos-fw-accept"
       ]);
+      services.udev.extraRules = mkIf ("${name}" == "smartctl") ''
+        SUBSYSTEM=="nvme", KERNEL=="nvme[0-9]*", GROUP="disk"
```

Comment:
I would be fine with a dedicated `rawio` group here. The `disk` group already offers something else.

Reply:
As an afterthought, it doesn't feel good to create a generic group `rawio` when configuring smartctl-exporter. I'll push a commit in a moment with it.

Reply:
I'm not convinced that `disk` is different, and I think that `rawio` would be misleading. TTBOMK the only interesting thing offered by top-level devices (as opposed to namespace devices, which are group `disk`) is creation and management of namespaces. I don't think that's "raw I/O" by any stretch. (Note that namespace devices do offer quering of smart status, but `smartctl --scan` will query the top-level ones. Arguably that's a sound choice, because smart status is per-device and not per-namespace)

Reply:
Should we just have a dedicated group like smartctl-exporter-nvme with granted permission created only if exporter is enabled? In that case, it'll be isolated to the problem we're solving here and it'll not mess with the generic groups.

Reply:
Well, we need to set (via an udev rule) the group owner on nvme devices to something. You can't set it to two different values (udev rules don't really support POSIX ACLs), so setting it to something smartctl-exporter-specific would conflict with anyone else trying to use this.

Given that, it might be useful to set this group unconditionally, so that people don't encounter very mysterious interactions between smartctl-exporter and rest of their configuration (either by it relying on what smartctl-exporter's setup does, or it being in conflict with that).

My own suggestion would be to, unconditionally, set this group to `disk`, but I expect many would disagree. I think that arguments against that stemming from different kinds of access have a very weak basis, for reasons I gave above.

Reply:
https://github.com/systemd/systemd/issues/26009#issuecomment-1377601680 is the distinction I would like to follow, and why it should not be part of `disk`.

And yes, making it something generic would be preferable to me as well.

Reply:
Well, but what about my response to that two comments below? We have a device where you can do _all_ the non-reading-bytes things, except for creating and destroying namespaces. This seems to be a very arbitrary boundary; similarly one could argue that access do the devices that contain partition tables should not be given to `disk`, because that's morally equivalent to creating namespaces (it's also repartitioning a device after a fashion).

I would be very happy with something other than `disk` _if we had a good definition of what `disk` should allow and stuck to it_. The "it should only allow access to byte reading and writing" approach is infeasible to follow for SATA, PATA, and probably other mass storage: they all also allow an opener of the device (and all its partitions) access to many of the commands related to smart status, readahead configuration, etc. via ioctls.

Reply:
I'm just opposed to mixing it with `disk`, as that stems from a convention that systemd dictates. I'm also opposed to making the group smartctl-exporter specific, as the exporter shouldn't have a special claim on the resouce in question.

So we are just looking for a proper name. We could narrow it down to just this single use case and go with `nvme` and possibly some descriptive suffix.

Reply:
I think `nvme` is an okay name for it and having it in a separate group wouldn't hurt and will not imperceptibly add permissions to existing services/users using the `disk` group. We should have it outside of the exporter in that case and unconditional, right?

Reply:
Unfortunately, I wasn't able to find a good place for the `nvme` group creation yet. Maybe any suggestions? Thank you.

Reply:
`nixosRules` in `services/hardware/udev.nix` possibly.

Reply:
> Well, we need to set (via an udev rule) the group owner on nvme devices to something. You can't set it to two different values (udev rules don't really support POSIX ACLs), so setting it to something smartctl-exporter-specific would conflict with anyone else trying to use this.

https://github.com/NixOS/nixpkgs/pull/333961 uses a smartctl_exporter specific group and POSIX ACL in the udev rule. The result is we can keep the base/standard `root:root` ownership of the NVMe device nodes while simultaneously giving smartctl_exporter access. I think this is a pretty clean solution, withouth side effects (AFAICT).